src/lib/Microsoft.Health.Events/EventHubProcessor/PartitionLockingBackgroundService.cs

```diff
@@ -0,0 +1,487 @@
+            _totalProcessorsRunning = activeProcessorCount;
+            _logger.LogTrace($"Active processor count: {activeProcessorCount}");
+
+            if (activeProcessorCount == 0)
+            {
+                throw new Exception("active processors is 0");
+            }
+
+            // divide the partitions among the active pods
+            // todo: handle cases where there is a remainder
+            // for example 8 partitions and 6 pods, which would require 4 pods to have 1 partition, and 2 pods to have 2 partitions
+            // for now, we will round down
+
+            _logger.LogTrace($"The Event Hub has {_eventHubPartitions.Length} partitions");
+
+            var suggestedPartitionsPerProcessor = Math.Floor((double)_eventHubPartitions.Length / activeProcessorCount);
+            _logger.LogTrace($"Suggested partitions per processor: {suggestedPartitionsPerProcessor}");
+
+            // find partitions that are not leased, and lease them
+            //
+            // todo: detect if this takes too long
+            // for example, if we acquire 2 partitions right away but need 2 more and cannot get them,
+            // should we renew the 2 that we have, or not? Alternatively we could quit after X minutes
+            _ownedPartitions.Clear();
+
+            int partitionIdIndex = 0;
+            while (_ownedPartitions.Count < suggestedPartitionsPerProcessor && partitionIdIndex < _eventHubPartitions.Length)
+            {
+                // attempt to claim a partition
+                var partitionId = int.Parse(_eventHubPartitions[partitionIdIndex]);
+                _logger.LogTrace($"Attempting to claim partition id: {partitionId}");
+
+                try
+                {
+                    await ClaimPartitionUsingBlob(partitionId);
+                }
+                catch (RequestFailedException ex)
+                {
+                    if (ex.ErrorCode == BlobErrorCode.LeaseAlreadyPresent)
+                    {
+                        _logger.LogTrace($"Failed to acquire partition {partitionId}");
+                        await Task.Delay(1000, ct);
+                        partitionIdIndex++;
+
+                        // reset if we have reached the end of the partition list but have not acquired enough partitions
+                        if (partitionId + 1 >= _eventHubPartitions.Length && _ownedPartitions.Count != suggestedPartitionsPerProcessor)
+                        {
+                            _logger.LogTrace("Not enough partitions are available for processing... Waiting");
+                            await Task.Delay(5000, ct);
+                            partitionIdIndex = 0;
+                            continue;
+                        }
+
+                        continue;
+                    }
+                    else
+                    {
+                        throw;
```

Comment:
Should we have a retry policy with some minor back off?  Trying to think if there are any transient exceptions that could get us into trouble here.  It looks like an exception here will kill the entire Execute block.   Have we verified this properly kills the process?